### PR TITLE
Improve the `multiIf` function performance when the type is Nullable.

### PR DESCRIPTION
**Pull request by KevinyhZou**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve the `multiIf` function performance when the type is Nullable.

#### Improve on nullable type
We find in `multiIf` function, when the return column type is `nullable`, then it would not use `executeInstructionsColumnar` method, which may decrease its performance. So we introduce a similar method `executeInstructionsColumnarForNullable` for nullable data type, to improve the function's performance in nullable case

Here, we test its performance below, the test table `test_tbl(d Nullable(Int64))` has about 300000000 rows data range from 1~300000000, and we do the query ` select count(1) from test_tbl where multiIf(d > 2, d-2, d > 1, d-1, d >0, d, 0) > 1 SETTINGS max_threads=1`, test for 3 times

before this pr's modification, test as below
```
1 row in set. Elapsed: 8.121 sec. Processed 300.00 million rows, 2.70 GB (36.94 million rows/s., 332.47 MB/s.)
1 row in set. Elapsed: 7.169 sec. Processed 300.00 million rows, 2.70 GB (41.84 million rows/s., 376.60 MB/s.)
1 row in set. Elapsed: 7.674 sec. Processed 300.00 million rows, 2.70 GB (39.09 million rows/s., 351.83 MB/s.)
```

after this pr's modification, test as below
```
1 row in set. Elapsed: 3.506 sec. Processed 300.00 million rows, 2.70 GB (85.56 million rows/s., 770.08 MB/s.)
1 row in set. Elapsed: 3.664 sec. Processed 300.00 million rows, 2.70 GB (81.88 million rows/s., 736.96 MB/s.)
1 row in set. Elapsed: 3.511 sec. Processed 300.00 million rows, 2.70 GB (85.46 million rows/s., 769.12 MB/s.)
```
### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 595798f4355f76c9218c829b859df3879c9b2b5c with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57745/595798f4355f76c9218c829b859df3879c9b2b5c/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7167113568">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57745/595798f4355f76c9218c829b859df3879c9b2b5c/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Functions/multiIf.cpp [hunk 1]
```diff
@@ -373,6 +382,47 @@ class FunctionMultiIf final : public FunctionIfBase
         }
     }
 
+    template<typename T, typename S>
+    static void executeInstructionsColumnarForNullable(std::vector<Instruction> & instructions, size_t rows, const MutableColumnPtr & res, const MutableColumnPtr & null_map)
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:386]: It is possible to reuse executeInstructionsColumnar by adding a template param <bool nullable> for it ?

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: src/Functions/multiIf.cpp [hunk 2]
```diff
@@ -238,7 +238,7 @@ class FunctionMultiIf final : public FunctionIfBase
         }
 
         const auto & settings = context->getSettingsRef();
-        const WhichDataType which(result_type);
+        const WhichDataType which(removeNullable(result_type));
         bool execute_multiif_columnar
             = settings.allow_execute_multiif_columnar && !contains_short && (which.isInt() || which.isUInt() || which.isFloat());
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:243]: It is possible to apply columnar execution to decimal types

File: src/Functions/multiIf.cpp [hunk 3]
```diff
@@ -414,27 +415,87 @@ class FunctionMultiIf final : public FunctionIfBase
...
+                    }
+                    insert_start_pos = row_i + 1;
+                }
+            }
         }
     }
 
     template <typename T, typename S>
-    static void executeInstructionsColumnar(std::vector<Instruction> & instructions, size_t rows, const MutableColumnPtr & res)
+    static void executeInstructionsColumnar(std::vector<Instruction> & instructions, size_t rows, const MutableColumnPtr & res, bool execute_by_memcpy)
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:462]: Need to prove that execute_by_memcpy = true has beffer performance than execute_by_memcpy  = false by adding a perf test.

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: src/Functions/multiIf.cpp [hunk 4]
```diff
@@ -373,18 +382,120 @@ class FunctionMultiIf final : public FunctionIfBase
...
+            for (size_t i = 0; i < instructions.size(); ++i)
+            {
+                data_cols[i] = assert_cast<const ColumnVector<T> *>(instructions[i].source.get());
+            }
+            for (size_t row_i = 0; row_i < rows; ++row_i)
+            {
+                S curr_insert = inserts[row_i];
+                if (row_i != rows -1 && curr_insert == inserts[row_i + 1])
+                {
+                    continue;
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:490]: may break compiler auto simd optimization.

File: src/Functions/multiIf.cpp (hunk 3 above)
Comment (KevinyhZou) [reply src/Functions/multiIf.cpp:462]: On the pr decription

File: src/Functions/multiIf.cpp [hunk 5]
```diff
@@ -373,18 +382,120 @@ class FunctionMultiIf final : public FunctionIfBase
...
+                }
+                else
+                {
+                    memcpy(res_data.data() + insert_start_pos,
+                        data_cols[curr_insert]->getData().data() + insert_start_pos, sizeof(T) * (row_i + 1 - insert_start_pos));
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:495]: why called getData every time ?  It is better to put  ColumnVector::getData() in vector instead of ColumnVector

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: src/Functions/multiIf.cpp (hunk 1 above)
Comment (KevinyhZou) [reply src/Functions/multiIf.cpp]: if reuse executeInstructionsColumnar, should add some if code, maybe not will for the `for` loop

File: src/Functions/multiIf.cpp (hunk 5 above)
Comment (KevinyhZou) [reply src/Functions/multiIf.cpp]: done

File: src/Functions/multiIf.cpp [hunk 6]
```diff
@@ -373,18 +385,115 @@ class FunctionMultiIf final : public FunctionIfBase
         }
     }
 
+    template<typename S>
+    static void calculateWhetherUseMemoryCopy(PaddedPODArray<S> & inserts, std::vector<Instruction> & instructions, std::optional<int> & instruction_use_memory_copy,
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:389]: calculateWhetherUseMemoryCopy -> calculateWhetherUseMemoryCopyIfNeeded

File: src/Core/Settings.h [hunk 7]
```diff
@@ -522,6 +522,7 @@ class IColumn;
     M(Bool, parsedatetime_parse_without_leading_zeros, true, "Formatters '%c', '%l' and '%k' in function 'parseDateTime()' parse months and hours without leading zeros.", 0) \
     M(Bool, formatdatetime_format_without_leading_zeros, false, "Formatters '%c', '%l' and '%k' in function 'formatDateTime()' print months and hours without leading zeros.", 0) \
     \
+    M(Float, skew_threshold_use_memcpy_execute_multiif_columnar, 0.9f, "The condition skew threshold to use memcpy while execute mulitiif columnar", 0) \
```
Comment (taiyang-li) [inline src/Core/Settings.h:525]: Looks skew_threshold_use_memcpy_in_multiif_columnar_execution is better

File: src/Functions/multiIf.cpp [hunk 8]
```diff
@@ -373,18 +385,115 @@ class FunctionMultiIf final : public FunctionIfBase
...
+        const MutableColumnPtr & null_map, std::optional<int> & instruction_use_memory_copy, double threshold_use_memcpy)
+    {
+        PaddedPODArray<S> inserts(rows, static_cast<S>(instructions.size()));
+        calculateInserts(instructions, rows, inserts);
+
+        PaddedPODArray<T> & res_data = assert_cast<ColumnVector<T> &>(*res).getData();
+        PaddedPODArray<UInt8> & null_map_data = assert_cast<ColumnUInt8 &>(*null_map).getData();
+        std::vector<const T*> data_cols(instructions.size());
+        std::vector<const UInt8 *> null_map_cols(instructions.size());
+        calculateWhetherUseMemoryCopy(inserts, instructions, instruction_use_memory_copy, threshold_use_memcpy);
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:416]: use_memory_copy is simpler

File: src/Functions/multiIf.cpp [hunk 9]
```diff
@@ -373,18 +385,115 @@ class FunctionMultiIf final : public FunctionIfBase
...
+                data_cols[i] = assert_cast<const ColumnVector<T> &>(*nullable_col->getNestedColumnPtr()).getData().data();
+            }
+            else
+            {
+                null_map_cols[i] = ColumnUInt8::create(rows)->getData().data();
+                data_cols[i] = assert_cast<const ColumnVector<T> &>(*instructions[i].source).getData().data();
+            }
+        }
+
+        if (!instruction_use_memory_copy.has_value())
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:439]: wrong logic. instruction_use_memory_copy alway has value here.

Comment (taiyang-li) [reply src/Functions/multiIf.cpp:439]:
```
if (!(*instruction_use_memory_copy)) 
{ 
   xxx
}

File: src/Functions/multiIf.cpp [hunk 10]
```diff
@@ -295,6 +305,8 @@ class FunctionMultiIf final : public FunctionIfBase
     }
 
 private:
+    mutable std::optional<int> instruction_use_memory_copy;
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:308]: use bool

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: src/Functions/multiIf.cpp [hunk 11]
```diff
@@ -373,18 +385,115 @@ class FunctionMultiIf final : public FunctionIfBase
...
+    {
+        if (!instruction_use_memory_copy.has_value())
+        {
+            std::vector<UInt64> instruction_insert_sizes(instructions.size());
+            for (size_t i = 0; i < inserts.size(); ++i)
+                instruction_insert_sizes[inserts[i]] += 1;
+            for (size_t i = 0; i < instruction_insert_sizes.size(); ++i)
+            {
+                if (!instructions[i].source_is_constant && instruction_insert_sizes[i] * 1.0 / inserts.size() >= threshold_use_memcpy)
+                    instruction_use_memory_copy.emplace(static_cast<int>(i));
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:400]: *instruction_use_memory_copy = true; break;

File: src/Functions/multiIf.cpp [hunk 12]
```diff
@@ -373,18 +385,115 @@ class FunctionMultiIf final : public FunctionIfBase
...
 
+    template<typename S>
+    static void calculateWhetherUseMemoryCopy(PaddedPODArray<S> & inserts, std::vector<Instruction> & instructions, std::optional<int> & instruction_use_memory_copy,
+        double threshold_use_memcpy)
+    {
+        if (!instruction_use_memory_copy.has_value())
+        {
+            std::vector<UInt64> instruction_insert_sizes(instructions.size());
+            for (size_t i = 0; i < inserts.size(); ++i)
+                instruction_insert_sizes[inserts[i]] += 1;
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:396]: use instruction_insert_sizes[inserts[i]]++

File: src/Functions/multiIf.cpp [hunk 13]
```diff
@@ -373,18 +385,115 @@ class FunctionMultiIf final : public FunctionIfBase
         }
     }
 
+    template<typename S>
+    static void calculateWhetherUseMemoryCopy(PaddedPODArray<S> & inserts, std::vector<Instruction> & instructions, std::optional<int> & instruction_use_memory_copy,
+        double threshold_use_memcpy)
+    {
+        if (!instruction_use_memory_copy.has_value())
+        {
+            std::vector<UInt64> instruction_insert_sizes(instructions.size());
```
Comment (taiyang-li) [inline src/Functions/multiIf.cpp:394]: Initialize all elements in vector to zero. 

instruction_insert_sizes(instructions.size(), 0);

**Review by taiyang-li — CHANGES_REQUESTED** [review]

**Comment by KevinyhZou** [comment]
It seems to use memcpy when the condition skew not improve the performance, so I have remove the memcpy copy logical, and I have reuse `executeInstructionsColumnar` to handle nullable case. @taiyang-li

**Comment by KevinyhZou** [comment]
cc @evillique

File: src/Functions/multiIf.cpp [hunk 14]
```diff
@@ -254,8 +253,12 @@ class FunctionMultiIf final : public FunctionIfBase
     if (which.is##TYPE()) \
     { \
         MutableColumnPtr res = ColumnVector<TYPE>::create(rows); \
-        executeInstructionsColumnar<TYPE, INDEX>(instructions, rows, res); \
-        return std::move(res); \
+        MutableColumnPtr null_map = ColumnUInt8::create(rows); \
```
Comment (evillique) [inline src/Functions/multiIf.cpp:256]: Do we need to create `null_map` irregardless of `result_type->isNullable()`?

Comment (KevinyhZou) [reply src/Functions/multiIf.cpp]: done

**Review by evillique — APPROVED** [review]

File: src/Functions/multiIf.cpp [hunk 15]
```diff
@@ -374,17 +378,54 @@ class FunctionMultiIf final : public FunctionIfBase
...
+                    {
+                        const ColumnPtr data_column = assert_cast<const ColumnConst &>(*instructions[i].source).getDataColumnPtr();
+                        nullable_col = assert_cast<const ColumnNullable *>(data_column.get());
+                    }
+                    null_map_cols[i] = assert_cast<const ColumnUInt8 &>(*nullable_col->getNullMapColumnPtr()).getData().data();
+                    data_cols[i] = assert_cast<const ColumnVector<T> &>(*nullable_col->getNestedColumnPtr()).getData().data();
+                }
+                else
+                {
+                    null_map_cols[i] = ColumnUInt8::create(rows)->getData().data();
```
Comment (evillique) [inline src/Functions/multiIf.cpp:418]: We should also probably reuse a single `ColumnUInt8` for different columns

Comment (KevinyhZou) [reply src/Functions/multiIf.cpp]: done @evillique

Comment (evillique) [reply src/Functions/multiIf.cpp]: Well it looks like if the first column was nullable then we would use some nullable filter on a non-nullable column

Comment (KevinyhZou) [reply src/Functions/multiIf.cpp]: Fixed . @evillique

**Comment by KevinyhZou** [comment]
cc @evillique  can you help to merge this pr?

**Pull request merged**